[CMake][DFT] target_link_library(myTarget PRIVATE MKL::onemkl_dft_XXX) does not add includes

# Summary

The oneMKL interface library supports linking the the DFT backends (to allow for compile-time dispatch) instead of linking to the main library itself and using the dynamic dispatch mechanism. There are examples of using oneMKL in this way in the examples.

However, the CMake in the examples does is not representative of oneMKL usage after installation.

# Observed behavior

Using the example of the MKLGPU backend, I'd expect to the following:
```
        target_link_libraries(libgromacs PRIVATE MKL::onemkl_dft_mklgpu)
```
to also include the MKL header files. e.g. I'd expect to able to `#include <oneapi/mkl/dft.hpp>`. However, without additional cmake, this header is not available.

# Expected behavior

I think the DFT backends targets should include the relevent headers as part of their interface.

# Version
Report oneMKL version: develop tip: [f5dc527](https://github.com/oneapi-src/oneMKL/commit/f5dc52762a042fbfb5f48f2e01b81463ae9b05f8)

# Environment
DPC++ 2024.0
Ubuntu20

# Steps to reproduce

CMake:
```
cmake_minimum_required(VERSION 3.26)
project(exampleProject)
add_executable(myapp main.cpp)
find_package(oneMKL REQUIRED)
# target_link_libraries(myapp PRIVATE MKL::onemkl)
target_link_libraries(myapp PRIVATE MKL::onemkl_dft_mklgpu)
```

C++
```
#include <oneapi/mkl/dft.hpp>
int main(){ return 0; }
```
Gives:
```
FAILED: CMakeFiles/myapp.dir/main.cpp.o
/opt/intel/oneapi/compiler/2024.0/bin/icpx   -fsycl -MD -MT CMakeFiles/myapp.dir/main.cpp.o -MF CMakeFiles/myapp.dir/main.cpp.o.d -o CMakeFiles/myapp.dir/main.cpp.o -c /home/hugh/tmp/example/main.cpp
/home/hugh/tmp/example/main.cpp:1:10: fatal error: 'oneapi/mkl/dft.hpp' file not found
    1 | #include <oneapi/mkl/dft.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~
1 error generated.
```

--- a/examples/dft/compile_time_dispatching/CMakeLists.txt
+++ b/examples/dft/compile_time_dispatching/CMakeLists.txt
@@ -30,19 +30,17 @@ foreach(dft_ct_source ${DFT_CT_SOURCES})
   add_executable(${EXAMPLE_NAME} ${dft_ct_source}.cpp)
   target_include_directories(${EXAMPLE_NAME}
       PUBLIC ${PROJECT_SOURCE_DIR}/examples/include
-      PUBLIC ${PROJECT_SOURCE_DIR}/include
       PUBLIC ${CMAKE_BINARY_DIR}/bin
   )
 
-if(domain STREQUAL "dft" AND ENABLE_MKLCPU_BACKEND AND ENABLE_CUFFT_BACKEND)
-  add_dependencies(${EXAMPLE_NAME} onemkl_${domain}_mklcpu onemkl_${domain}_cufft)
-  list(APPEND ONEMKL_LIBRARIES_${domain} onemkl_${domain}_mklcpu onemkl_${domain}_cufft)
-endif()
+  if(domain STREQUAL "dft" AND ENABLE_MKLCPU_BACKEND AND ENABLE_CUFFT_BACKEND)
+    add_dependencies(${EXAMPLE_NAME} onemkl_${domain}_mklcpu onemkl_${domain}_cufft)
+    list(APPEND ONEMKL_LIBRARIES_${domain} onemkl_${domain}_mklcpu onemkl_${domain}_cufft)
+  endif()
 
-target_link_libraries(${EXAMPLE_NAME} PUBLIC
-  ${ONEMKL_LIBRARIES_${domain}}
-  ONEMKL::SYCL::SYCL
-  onemkl_warnings
+  target_link_libraries(${EXAMPLE_NAME} PUBLIC
+    ${ONEMKL_LIBRARIES_${domain}}
+    onemkl_warnings
   )
 
   # Register example as ctest

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -32,6 +32,13 @@ set(ENABLE_PORTBLAS_BACKEND_NVIDIA_GPU OFF CACHE INTERNAL "")
 # store path to CMAKE_CURRENT_BINARY_DIR to use it later (makes FetchContent_Declare workable)
 set(ONEMKL_GENERATED_INCLUDE_PATH ${CMAKE_CURRENT_BINARY_DIR})
 
+
+set(ONEMKL_INTERFACE_INCLUDE_DIRS
+  $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>
+  $<INSTALL_INTERFACE:include>
+)
+
 # Build loader and backends for each domain
 add_custom_target(onemkl_backend_libs)
 foreach(domain ${TARGET_DOMAINS})
@@ -60,9 +67,7 @@ if(BUILD_SHARED_LIBS)
   add_dependencies(onemkl onemkl_backend_libs)
 
   target_include_directories(onemkl
-    PUBLIC $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
-           $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>
-           $<INSTALL_INTERFACE:include>
+    PUBLIC ${ONEMKL_INTERFACE_INCLUDE_DIRS}
   )
   set_target_properties(onemkl PROPERTIES
     SOVERSION ${PROJECT_VERSION_MAJOR}

--- a/src/dft/backends/cufft/CMakeLists.txt
+++ b/src/dft/backends/cufft/CMakeLists.txt
@@ -34,6 +34,9 @@ add_dependencies(onemkl_backend_libs_dft ${LIB_NAME})
 target_include_directories(${LIB_OBJ}
   PUBLIC ${ONEMKL_INTERFACE_INCLUDE_DIRS}
 )
+target_include_directories(${LIB_NAME}
+  PUBLIC ${ONEMKL_INTERFACE_INCLUDE_DIRS}
+)
 target_include_directories(${LIB_OBJ}
   PRIVATE ${PROJECT_SOURCE_DIR}/src
           ${CMAKE_BINARY_DIR}/bin

--- a/src/dft/backends/cufft/CMakeLists.txt
+++ b/src/dft/backends/cufft/CMakeLists.txt
@@ -32,8 +32,10 @@ add_library(${LIB_OBJ} OBJECT
 add_dependencies(onemkl_backend_libs_dft ${LIB_NAME})
 
 target_include_directories(${LIB_OBJ}
-  PRIVATE ${PROJECT_SOURCE_DIR}/include
-          ${PROJECT_SOURCE_DIR}/src
+  PUBLIC ${ONEMKL_INTERFACE_INCLUDE_DIRS}
+)
+target_include_directories(${LIB_OBJ}
+  PRIVATE ${PROJECT_SOURCE_DIR}/src
           ${CMAKE_BINARY_DIR}/bin
           ${ONEMKL_GENERATED_INCLUDE_PATH}
 )

--- a/src/dft/backends/mklcpu/CMakeLists.txt
+++ b/src/dft/backends/mklcpu/CMakeLists.txt
@@ -33,8 +33,10 @@ add_library(${LIB_OBJ} OBJECT
 add_dependencies(onemkl_backend_libs_dft ${LIB_NAME})
 
 target_include_directories(${LIB_OBJ}
-  PRIVATE ${PROJECT_SOURCE_DIR}/include
-          ${PROJECT_SOURCE_DIR}/src
+  PUBLIC ${ONEMKL_INTERFACE_INCLUDE_DIRS}
+)
+target_include_directories(${LIB_OBJ}
+  PRIVATE ${PROJECT_SOURCE_DIR}/src
           ${CMAKE_BINARY_DIR}/bin
           ${ONEMKL_GENERATED_INCLUDE_PATH}
 )

--- a/src/dft/backends/mklcpu/CMakeLists.txt
+++ b/src/dft/backends/mklcpu/CMakeLists.txt
@@ -35,6 +35,10 @@ add_dependencies(onemkl_backend_libs_dft ${LIB_NAME})
 target_include_directories(${LIB_OBJ}
   PUBLIC ${ONEMKL_INTERFACE_INCLUDE_DIRS}
 )
+target_include_directories(${LIB_NAME}
+  PUBLIC ${ONEMKL_INTERFACE_INCLUDE_DIRS}
+)
+
 target_include_directories(${LIB_OBJ}
   PRIVATE ${PROJECT_SOURCE_DIR}/src
           ${CMAKE_BINARY_DIR}/bin

--- a/src/dft/backends/mklgpu/CMakeLists.txt
+++ b/src/dft/backends/mklgpu/CMakeLists.txt
@@ -33,8 +33,10 @@ add_library(${LIB_OBJ} OBJECT
 add_dependencies(onemkl_backend_libs_dft ${LIB_NAME})
 
 target_include_directories(${LIB_OBJ}
-  PRIVATE ${PROJECT_SOURCE_DIR}/include
-          ${PROJECT_SOURCE_DIR}/src
+  PUBLIC ${ONEMKL_INTERFACE_INCLUDE_DIRS}
+)
+target_include_directories(${LIB_OBJ}
+  PRIVATE ${PROJECT_SOURCE_DIR}/src
           ${CMAKE_BINARY_DIR}/bin
           ${ONEMKL_GENERATED_INCLUDE_PATH}
 )

--- a/src/dft/backends/mklgpu/CMakeLists.txt
+++ b/src/dft/backends/mklgpu/CMakeLists.txt
@@ -35,6 +35,9 @@ add_dependencies(onemkl_backend_libs_dft ${LIB_NAME})
 target_include_directories(${LIB_OBJ}
   PUBLIC ${ONEMKL_INTERFACE_INCLUDE_DIRS}
 )
+target_include_directories(${LIB_NAME}
+  PUBLIC ${ONEMKL_INTERFACE_INCLUDE_DIRS}
+)
 target_include_directories(${LIB_OBJ}
   PRIVATE ${PROJECT_SOURCE_DIR}/src
           ${CMAKE_BINARY_DIR}/bin

--- a/src/dft/backends/portfft/CMakeLists.txt
+++ b/src/dft/backends/portfft/CMakeLists.txt
@@ -93,6 +93,9 @@ target_link_libraries(${LIB_OBJ} PRIVATE onemkl_warnings)
 target_include_directories(${LIB_OBJ}
   PUBLIC ${ONEMKL_INTERFACE_INCLUDE_DIRS}
 )
+target_include_directories(${LIB_NAME}
+  PUBLIC ${ONEMKL_INTERFACE_INCLUDE_DIRS}
+)
 target_include_directories(${LIB_OBJ}
   PRIVATE ${PROJECT_SOURCE_DIR}/src
           ${CMAKE_BINARY_DIR}/bin

--- a/src/dft/backends/portfft/CMakeLists.txt
+++ b/src/dft/backends/portfft/CMakeLists.txt
@@ -91,8 +91,10 @@ endif()
 target_link_libraries(${LIB_OBJ} PRIVATE onemkl_warnings)
 
 target_include_directories(${LIB_OBJ}
-  PRIVATE ${PROJECT_SOURCE_DIR}/include
-          ${PROJECT_SOURCE_DIR}/src
+  PUBLIC ${ONEMKL_INTERFACE_INCLUDE_DIRS}
+)
+target_include_directories(${LIB_OBJ}
+  PRIVATE ${PROJECT_SOURCE_DIR}/src
           ${CMAKE_BINARY_DIR}/bin
 )
 

--- a/src/dft/backends/rocfft/CMakeLists.txt
+++ b/src/dft/backends/rocfft/CMakeLists.txt
@@ -34,6 +34,9 @@ add_dependencies(onemkl_backend_libs_dft ${LIB_NAME})
 target_include_directories(${LIB_OBJ}
   PUBLIC ${ONEMKL_INTERFACE_INCLUDE_DIRS}
 )
+target_include_directories(${LIB_NAME}
+  PUBLIC ${ONEMKL_INTERFACE_INCLUDE_DIRS}
+)
 target_include_directories(${LIB_OBJ}
   PRIVATE ${PROJECT_SOURCE_DIR}/src
           ${CMAKE_BINARY_DIR}/bin

--- a/src/dft/backends/rocfft/CMakeLists.txt
+++ b/src/dft/backends/rocfft/CMakeLists.txt
@@ -32,8 +32,10 @@ add_library(${LIB_OBJ} OBJECT
 add_dependencies(onemkl_backend_libs_dft ${LIB_NAME})
 
 target_include_directories(${LIB_OBJ}
-  PRIVATE ${PROJECT_SOURCE_DIR}/include
-          ${PROJECT_SOURCE_DIR}/src
+  PUBLIC ${ONEMKL_INTERFACE_INCLUDE_DIRS}
+)
+target_include_directories(${LIB_OBJ}
+  PRIVATE ${PROJECT_SOURCE_DIR}/src
           ${CMAKE_BINARY_DIR}/bin
           ${ONEMKL_GENERATED_INCLUDE_PATH}
 )

